src/fermilib/utils/_graph.py

```diff
@@ -0,0 +1,222 @@
+
+        Args:
+            node_id1(int): Index of first node on edge
+            node_id2(int): Index of second node on edge
+        """
+        if node_id1 >= self.node_count() or node_id2 >= self.node_count():
+            raise IndexError("Node ID out of range for graph")
+
+        self.neighbors[node_id1].add(self.node_uids[node_id2])
+        self.neighbors[node_id2].add(self.node_uids[node_id1])
+
+    def remove_edge(self, node_id1, node_id2):
+        """Remove an edge between node1 and node2
+
+        Args:
+            node_id1(int): Index of first node on edge
+            node_id2(int): Index of second node on edge
+        """
+        if node_id1 >= self.node_count() or node_id2 >= self.node_count():
+            raise IndexError("Node ID out of range for graph")
+
+        self.neighbors[node_id1].remove(self.node_uids[node_id2])
+        self.neighbors[node_id2].remove(self.node_uids[node_id1])
+
+    def find_index(self, value, starting_node=0):
+        """Find the index of the first node that matches value in a BFS
+
+        Performs a breadth-first search of the graph starting at node index
+        starting_node.  Returns the index or None if no match is found
+
+        Args:
+            value(Node Value) - Value to match against in the graph
+            starting_node(int) - Node index to start search from
+        """
+        if starting_node > self.node_count():
+            raise IndexError("Node ID out of range.")
+
+        node_queue = queue.LifoQueue()
+        node_queue.put(starting_node)
+        visited = [starting_node]
+        while not node_queue.empty():
+            next_id = node_queue.get()
+            if (self.nodes[next_id].value == value):
+                return next_id  # Success
+            for uid in self.neighbors[next_id]:
+                if (self.uid_to_index[uid] not in visited):
+                    node_queue.put(self.uid_to_index[uid])
+                    visited += [self.uid_to_index[uid]]
+        return None
+
+    def is_adjacent(self, node_id1, node_id2):
+        """Test for adjacency between node1 and node2
+
+         Args:
+            node_id1(int): Index of first node
+            node_id2(int): Index of second node
+        """
+        if node_id1 >= self.node_count() or node_id2 >= self.node_count():
+            raise IndexError("Node ID out of range for graph")
+
+        return (self.node_uids[node_id1] in self.neighbors[node_id2])
+
+    def get_neighbors(self, node_id):
+        """Return list of neighbors of the specified node
+
+        Args:
+            node_id: Index of node to examine the neighbors of
+
+        Returns(list): List of current node IDs that are neighbors of node_id.
+        """
+        if node_id >= self.node_count():
+            raise IndexError("Node ID out of range for graph")
+        return [self.uid_to_index[i] for i in self.neighbors[node_id]]
+
+    def shortest_path(self, node_id1, node_id2):
```

Comment:
Looks like this function has some logics in common with find_index(), after all, both of them use BFS - it'll be nice to abstract that out and share the logic between these two functions.
(Could be a future refactor.)